ios/FluentUI/Bottom Sheet/BottomSheetController.swift

```diff
@@ -0,0 +1,376 @@
+                move(to: .collapsed, velocity: velocity)
+            } else {
+                move(to: .expanded, velocity: velocity)
+            }
+        }
+    }
+
+    private func move(to targetExpansionState: BottomSheetExpansionState, animated: Bool = true, velocity: CGFloat = 0.0) {
+        if animated {
+            animate(to: targetExpansionState, velocity: velocity)
+        } else {
+            bottomSheetOffsetConstraint.constant = -(targetExpansionState == .expanded ? expandedOffsetFromBottom : collapsedOffsetFromBottom)
+            view.setNeedsLayout()
+
+            switch targetExpansionState {
+            case .expanded:
+                delegate?.bottomSheetControllerDidExpand?(self)
+            case .collapsed:
+                delegate?.bottomSheetControllerDidCollapse?(self)
+            }
+        }
+    }
+
+    private func animate(to targetExpansionState: BottomSheetExpansionState, velocity: CGFloat = 0.0) {
+        let targetOffsetFromBottom = targetExpansionState == .expanded ? expandedOffsetFromBottom : collapsedOffsetFromBottom
+        let distanceToGo = abs(currentOffsetFromBottom - targetOffsetFromBottom)
+        let springVelocity = min(abs(velocity / distanceToGo), Constants.Spring.maxInitialVelocity)
+        let damping: CGFloat = abs(velocity) > Constants.Spring.flickVelocityThreshold
+            ? Constants.Spring.oscillatingDampingRatio
+            : Constants.Spring.defaultDampingRatio
+
+        let springParams = UISpringTimingParameters(dampingRatio: damping, initialVelocity: CGVector(dx: 0.0, dy: springVelocity))
+        translationAnimator = UIViewPropertyAnimator(duration: Constants.Spring.animationDuration, timingParameters: springParams)
+
+        view.layoutIfNeeded()
+        bottomSheetOffsetConstraint.constant = -targetOffsetFromBottom
+        translationAnimator?.addAnimations {
+            self.view.layoutIfNeeded()
+        }
+
+        translationAnimator?.addCompletion({ finalPosition in
+            if finalPosition == .end {
+                switch targetExpansionState {
+                case .expanded:
+                    self.delegate?.bottomSheetControllerDidExpand?(self)
+                case .collapsed:
+                    self.delegate?.bottomSheetControllerDidCollapse?(self)
+                }
+            }
+        })
+        translationAnimator?.startAnimation()
+    }
+
+    private func stopAnimationIfNeeded() {
+        guard let animator = translationAnimator else {
+            return
+        }
+
+        if animator.isRunning {
+            animator.stopAnimation(true)
+
+            // The AutoLayout constant doesn't animate, so we need to set it to whatever it should be
+            // based on the frame calculated during the interrupted animation
+            let offsetFromBottom = view.frame.height - bottomSheetView.frame.origin.y
+            bottomSheetOffsetConstraint.constant = -offsetFromBottom
+            view.setNeedsLayout()
```

Comment:
Similar comment - what happens if we don't have it. Does adjusting the constant not already kick off a layout pass?

If we need layout to happen deterministically before something else happens then we would need layoutIfNeeded?